File: astar.py
import copy
import heapq

class Node:
    def __init__(self, mat, x, y, gscore, parent=None):
        self.mat = mat
        self.x = x
        self.y = y
        self.gscore = gscore
        self.parent = parent
        self.hscore = self.calculate_cost()

    def calculate_cost(self):
        count = 0
        for i in range(3):
            for j in range(3):
                if self.mat[i][j] != 0 and self.mat[i][j] != final[i][j]:
                    count += 1
        return count

    def __lt__(self, other):
        return (self.hscore + self.gscore) < (other.hscore + other.gscore)

def print_matrix(mat):
    for row in mat:
        print(*row)

def is_safe(x, y):
    return 0 <= x < 3 and 0 <= y < 3

def print_path(root):
    if not root:
        return
    print_path(root.parent)
    print_matrix(root.mat)
    print("hscore:", root.hscore)
    print("gscore:", root.gscore)
    print("fscore:", root.hscore + root.gscore, "\n")

def solve(initial, x, y, final):
    pq = []
    root = Node(initial, x, y, 0)
    heapq.heappush(pq, root)
    while pq:
        min_node = heapq.heappop(pq)
        if min_node.hscore == 0:
            print_path(min_node)
            return
        for i, j in [(1, 0), (0, -1), (-1, 0), (0, 1)]:
            if is_safe(min_node.x + i, min_node.y + j):
                new_mat = copy.deepcopy(min_node.mat)
                new_mat[min_node.x][min_node.y], new_mat[min_node.x + i][min_node.y + j] = new_mat[min_node.x + i][min_node.y + j], new_mat[min_node.x][min_node.y]
                child = Node(new_mat, min_node.x + i, min_node.y + j, min_node.gscore + 1, min_node)
                heapq.heappush(pq, child)

initial = [[0]*3 for _ in range(3)]
x, y = 0, 0

print("\nEnter Initial Block Structure\nEnter 0 for blank space:")
for i in range(3):
    for j in range(3):
        initial[i][j] = int(input("Row {} Column {} Element = ".format(i + 1, j + 1)))
        if initial[i][j] == 0:
            x, y = i, j

final = [[0]*3 for _ in range(3)]
print("\nEnter Final Block Structure\nEnter 0 for blank space:")
for i in range(3):
    for j in range(3):
        final[i][j] = int(input("Row {} Column {} Element = ".format(i + 1, j + 1)))

print("\nInitial Structure:")
print_matrix(initial)
print("\nFinal Structure:")
print_matrix(final)
print("\n\nThis is the solution using A* Algorithm:\n")
solve(initial, x, y, final)

File: astar.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node{
Node *parent;
int mat[3][3];
int x, y;
int hscore;
int gscore;
};
void printMatrix(int mat[3][3]){
for (int i = 0; i < 3; i++){
for (int j = 0; j < 3; j++)
cout << mat[i][j] << " ";
cout << "\n";
}
}
Node *newNode(int mat[3][3], int x, int y, int newX, int newY, int
gscore, Node *parent){
Node *node = new Node;
node->parent = parent;
memcpy(node->mat, mat, sizeof node->mat);
swap(node->mat[x][y], node->mat[newX][newY]);
node->hscore = INT_MAX;
node->gscore = gscore;
node->x = newX;
node->y = newY;
return node;
}
int row[] = {1, 0, -1, 0};
int col[] = {0, -1, 0, 1};
int calculateCost(int initial[3][3], int final[3][3]){
int count = 0;
for (int i = 0; i < 3; i++)
for (int j = 0; j < 3; j++)
if (initial[i][j] && initial[i][j] != final[i][j])
count++;
return count;
}


int isSafe(int x, int y){
return (x >= 0 && x < 3 && y >= 0 && y < 3);
}
void printPath(Node *root){
if (root == NULL)
return;
printPath(root->parent);
printMatrix(root->mat);
cout << "hscore: " << root->hscore << "\ngscore: " << root->gscore <<"\nfscore: " << root->hscore + root->gscore << "\n\n";
}
struct comp{
bool operator()(const Node *lhs, const Node *rhs) const{
return (lhs->hscore + lhs->gscore) > (rhs->hscore + rhs->gscore);
}
};
void solve(int initial[3][3], int x, int y, int final[3][3]){
priority_queue<Node *, std::vector<Node *>, comp> pq;
Node *root = newNode(initial, x, y, x, y, 0, NULL);
root->hscore = calculateCost(initial, final);
pq.push(root);
while (!pq.empty()){
Node *min = pq.top();
pq.pop();
if (min->hscore == 0){
printPath(min);
return;
}
for (int i = 0; i < 4; i++){
if (isSafe(min->x + row[i], min->y + col[i])){
Node *child = newNode(min->mat, min->x,
min->y, min->x + row[i],
min->y + col[i],
min->gscore + 1, min);
child->hscore = calculateCost(child->mat, final);
pq.push(child);
}
}
}
}
int main()
{
int initial[3][3];
int x, y;

cout << "\n\nEnter Initial Block Structure\n\nEnter 0 for blank space:\n\n";
for (int i = 0; i < 3; i++){
for (int j = 0; j < 3; j++){
cout << "Row " << i + 1 << " Column " << j + 1 << "Element = ";

cin >> initial[i][j];
if (initial[i][j] == 0){
x = i;
y = j;
}
}
}
int final[3][3];
cout << "\n\nEnter Final Block Structure\n\nEnter 0 for blank space:\n\n";
for (int i = 0; i < 3; i++){
for (int j = 0; j < 3; j++){
cout << "Row " << i + 1 << " Column " << j + 1 << " Element = ";

cin >> final[i][j];
}
}
cout << "\nInitial Structure:\n\n";
printMatrix(initial);
cout << "\nFinal Structure:\n\n";
printMatrix(final);
cout << "\n\nThis is the solution using A * Algorithm:\n\n";
solve(initial, x, y, final);
return 0;
}